# iscbtutorial

### ISCBTutorial23: Immune cell profiling from single-cell RNA with R

### Team

* Meghana Kshirsagar
* Gauri Vaidya
* Yang Ye


### Abstract 
Over the past couple of decades, immunotherapy treatments have been widely adopted as an alternative treatment for a variety of cancers. The study of tumour microenvironment of immune cells such as macrophages, T cells and B cells amongst others can help to unravel the mystery of differential outcomes to immunotherapy treatments. Gene expression profiling can help to identify the patterns of genes expressed in major immune cells amongst cohorts of patients at different stages of cancer to generate new biological hypotheses. Statistical approaches can facilitate the identification of highly variable genes and their expression in immune cells by performing analysis of scRNA sequencing data. The tutorial will be divided in three parts; comparing the popular annotation tools, applying dimensionality reduction techniques to obtain multi-stage downstreaming of scRNA data and building patient profiles from immune cell populations and subpopulations. Throughout the tutorial we will follow the seurat pipeline version 4.0.
 
### Pre-requisites
A basic understanding of R syntax would be helpful, but not required.
 
 
### _R_ / _Bioconductor_ packages used
  - Seurat
  - tidyverse
  - ggplot2
  - scCustomize
  - gridExtra
- DESeq2
- celldex
- ggpubr
- SingleR
- scMRMA
- SeuratWrappers
- Nebulosa
- dittoSeq
- harmony
- cowplot
- viridis


### Installation

You can access the tutorial from [github](https://meghanakshirsagar.github.io/icsbtutorial/) 

Slides are [online](https://docs.google.com/presentation/d/1VUcKvvb_nZAXyVBwGpQ_gndO5uLoz2UqtWHVtwbMWxg/edit#slide=id.p3)
